File: pc/McController.Core/Config/ConfigStore.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.Json.Serialization;

namespace McController.Core.Config;

/// <summary>
/// JSON-backed config persistence. Loads on startup, saves on demand
/// (debounced from the UI to avoid high-frequency disk writes).
///
/// Handles the pre-profile flat config layout (single camera/movement
/// at the root) by wrapping it into a one-element profile list during
/// load.
/// </summary>
public static class ConfigStore
{
    private static readonly JsonSerializerOptions s_options = new()
    {
        WriteIndented = true,
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        Converters = { new JsonStringEnumConverter(JsonNamingPolicy.CamelCase) },
    };

    public static ServerConfig LoadOrDefault(string path)
    {
        try
        {
            if (!File.Exists(path)) return new ServerConfig();
            var json = File.ReadAllText(path);
            var migrated = MigrateIfNeeded(json);
            return JsonSerializer.Deserialize<ServerConfig>(migrated, s_options) ?? new ServerConfig();
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"[ConfigStore] failed to load {path}: {ex.Message}. Using defaults.");
            return new ServerConfig();
        }
    }

    public static void Save(string path, ServerConfig config)
    {
        var json = JsonSerializer.Serialize(config, s_options);
        var dir = Path.GetDirectoryName(Path.GetFullPath(path));
        if (!string.IsNullOrEmpty(dir)) Directory.CreateDirectory(dir);
        File.WriteAllText(path, json);
    }

    /// <summary>
    /// Pre-profile configs had <c>camera</c>/<c>movement</c> at the top
    /// level and no <c>profiles</c> array. Wrap them into a single
    /// default profile so the rest of deserialization sees the new shape.
    /// New-shape configs pass through unchanged.
    /// </summary>
    private static string MigrateIfNeeded(string json)
    {
        JsonNode? root;
        try { root = JsonNode.Parse(json); }
        catch { return json; }
        if (root is not JsonObject obj) return json;
        if (obj.ContainsKey("profiles")) return json;
        var camera = obj["camera"];
        var movement = obj["movement"];
        if (camera is null && movement is null) return json;
        var profile = new JsonObject
        {
            ["id"] = "default",
            ["name"] = "默认",
            ["camera"] = camera?.DeepClone() ?? new JsonObject(),
            ["movement"] = movement?.DeepClone() ?? new JsonObject(),
        };
        obj["profiles"] = new JsonArray(profile);
        obj["activeProfileId"] = "default";
        obj.Remove("camera");
        obj.Remove("movement");
        return obj.ToJsonString();
    }
}

File: mac/McController/Core/Config/ConfigStore.swift
import Foundation

/// JSON-backed config persistence with one-shot migration for the
/// pre-profile flat layout (camera/movement at the root). Mirrors
/// `ConfigStore.cs` on the PC side but writes to
/// `~/Library/Application Support/McController/config.json` per
/// Apple's storage guidelines.
enum ConfigStore {

    /// Canonical location: `~/Library/Application Support/McController/config.json`.
    /// Created on demand. Falling back to NSHomeDirectory means the
    /// caller never crashes if `FileManager.default` returns nil for
    /// the app-support search.
    static func defaultConfigURL() -> URL {
        applicationSupportDirectory().appendingPathComponent("config.json", isDirectory: false)
    }

    static func applicationSupportDirectory() -> URL {
        let fm = FileManager.default
        let base = (try? fm.url(
            for: .applicationSupportDirectory,
            in: .userDomainMask,
            appropriateFor: nil,
            create: true))
            ?? URL(fileURLWithPath: NSHomeDirectory()).appendingPathComponent("Library/Application Support")
        let dir = base.appendingPathComponent("McController", isDirectory: true)
        try? fm.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    static func loadOrDefault(at url: URL) -> ServerConfig {
        do {
            guard FileManager.default.fileExists(atPath: url.path) else {
                return ServerConfig()
            }
            let data = try Data(contentsOf: url)
            let migrated = migrateIfNeeded(data)
            let decoder = JSONDecoder()
            return try decoder.decode(ServerConfig.self, from: migrated)
        } catch {
            NSLog("[ConfigStore] load failed: %@. Using defaults.", String(describing: error))
            return ServerConfig()
        }
    }

    static func save(_ config: ServerConfig, to url: URL) throws {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys, .withoutEscapingSlashes]
        let data = try encoder.encode(config)
        let dir = url.deletingLastPathComponent()
        try FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        try data.write(to: url, options: .atomic)
    }

    /// Pre-profile configs had `camera`/`movement` at the top level and
    /// no `profiles` array. Wrap them into a single default profile so
    /// the rest of deserialization sees the new shape. New-shape
    /// configs pass through unchanged.
    private static func migrateIfNeeded(_ data: Data) -> Data {
        guard
            let any = try? JSONSerialization.jsonObject(with: data, options: []),
            var obj = any as? [String: Any]
        else { return data }

        if obj["profiles"] != nil { return data }
        let camera = obj["camera"]
        let movement = obj["movement"]
        if camera == nil && movement == nil { return data }

        let profile: [String: Any] = [
            "id": "default",
            "name": "默认",
            "camera": camera ?? [:],
            "movement": movement ?? [:],
        ]
        // Some legacy configs spelled cameraConfig / movementConfig under
        // CamelCase variations; pass them through whatever shape they're in.
        obj["profiles"] = [profile]
        obj["activeProfileId"] = "default"
        obj.removeValue(forKey: "camera")
        obj.removeValue(forKey: "movement")
        return (try? JSONSerialization.data(withJSONObject: obj, options: [])) ?? data
    }
}
